### PR TITLE
Mf 10 employer choose course

### PR DESCRIPTION
**Pull request by robinoconnor**

### CODE REVIEW

File: src/SFA.DAS.Reservations.Web/Controllers/ReservationsController.cs [hunk 1]
```diff
@@ -61,9 +61,19 @@ public async Task<IActionResult> PostApprenticeshipTraining(ReservationsRouteMod
             StartDateModel startDateModel = null;
             if (!string.IsNullOrWhiteSpace(formModel.TrainingStartDate))
                 startDateModel = JsonConvert.DeserializeObject<StartDateModel>(formModel.TrainingStartDate);
+
             Course course = null;
-            if (!string.IsNullOrWhiteSpace(formModel.SelectedCourse))
-                course = JsonConvert.DeserializeObject<Course>(formModel.SelectedCourse);
+
```
Comment (scottywakefield) [inline src/SFA.DAS.Reservations.Web/Controllers/ReservationsController.cs]: again here just assign the course id and let the handler do the heavy lifting, no need to look up the course here.

Comment (robinoconnor) [reply src/SFA.DAS.Reservations.Web/Controllers/ReservationsController.cs]: Moved the logic to a new handler

File: src/SFA.DAS.Reservations.Web/Controllers/EmployerReservationsController.cs [hunk 2]
```diff
@@ -1,17 +1,128 @@
...
+                ReservationId = id,
+                Courses = courseViewModels
+            };
+
+            return View(viewModel);
+        }
+
+        [HttpPost]
+        [Route("{id}/SelectCourse")]
+        public async Task<IActionResult> PostSelectCourse(ReservationsRouteModel routeModel, string selectedCourseId)
```
[the remark this answers is not in the document]
Comment (robinoconnor) [reply src/SFA.DAS.Reservations.Web/Controllers/EmployerReservationsController.cs]: Moved the logic to a new handler

File: src/SFA.DAS.Reservations.Application/Reservations/Commands/CacheReservationCourse/CacheReservationCourseCommand.cs [hunk 3]
```diff
@@ -0,0 +1,12 @@
+﻿using System;
+using MediatR;
+
+namespace SFA.DAS.Reservations.Application.Reservations.Commands.CacheReservationCourse
+{
+    public class CacheReservationCourseCommand: IRequest<Unit>
+    {
+        public Guid? Id { get; set; }
```
Comment (dashton82) [inline src/SFA.DAS.Reservations.Application/Reservations/Commands/CacheReservationCourse/CacheReservationCourseCommand.cs]: Can be non-nullable

File: src/SFA.DAS.Reservations.Application/Reservations/Commands/CacheReservationEmployer/CacheReservationEmployerCommandValidator.cs [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+namespace SFA.DAS.Reservations.Application.Reservations.Commands.CacheReservationEmployer
+{
+    public class CacheReservationEmployerCommandValidator : IValidator<CacheReservationEmployerCommand>
+    {
+        public Task<ValidationResult> ValidateAsync(CacheReservationEmployerCommand command)
+        {
+            var result = new ValidationResult();
+
+            // id
```
Comment (dashton82) [inline src/SFA.DAS.Reservations.Application/Reservations/Commands/CacheReservationEmployer/CacheReservationEmployerCommandValidator.cs]: Remove the comments please

File: src/SFA.DAS.Reservations.Application/Reservations/Commands/CacheReservationStartDate/CacheReservationStartDateCommandHandler.cs [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+            {
+                throw new ValidationException(
+                    new ValidationResult("The following parameters have failed validation", queryValidationResult.ErrorList), null, null);
+            }
+
+            var cachedReservation = await _cacheStorageService.RetrieveFromCache<CachedReservation>(command.Id.ToString());
+
+            if (cachedReservation == null)
+            {
+                throw new Exception("No reservation was found with that Id");
```
Comment (dashton82) [inline src/SFA.DAS.Reservations.Application/Reservations/Commands/CacheReservationStartDate/CacheReservationStartDateCommandHandler.cs]: Shouldnt this be the newly created exception for Reservation not found? 
```suggestion
                throw new CachedReservationNotFoundException("No reservation was found with that Id");
```

File: src/SFA.DAS.Reservations.Application/Reservations/Commands/CacheReservationStartDate/CacheReservationStartDateCommandValidator.cs [hunk 6]
```diff
@@ -2,25 +2,26 @@
...
-    public class CacheCreateReservationCommandValidator : IValidator<CacheCreateReservationCommand>
+    public class CacheReservationStartDateCommandValidator : IValidator<CacheReservationStartDateCommand>
     {
-        public Task<ValidationResult> ValidateAsync(CacheCreateReservationCommand command)
+        public Task<ValidationResult> ValidateAsync(CacheReservationStartDateCommand command)
         {
             var result = new ValidationResult();
 
-            if (command.AccountId == default(long))
+            // id
```
Comment (dashton82) [inline src/SFA.DAS.Reservations.Application/Reservations/Commands/CacheReservationStartDate/CacheReservationStartDateCommandValidator.cs]: Remove comments please

File: src/SFA.DAS.Reservations.Web/Controllers/EmployerReservationsController.cs [hunk 7]
```diff
@@ -1,17 +1,125 @@
...
+                });
+            }
+            catch (ValidationException e)
+            {
+                foreach (var member in e.ValidationResult.MemberNames)
+                {
+                    ModelState.AddModelError(member.Split('|')[0], member.Split('|')[1]);
+                }
+
+                var getCoursesResponse = await _mediator.Send(new GetCoursesQuery());
```
Comment (dashton82) [inline src/SFA.DAS.Reservations.Web/Controllers/EmployerReservationsController.cs]: Might need to add some extra handling around this for if the courses dont come back

File: src/SFA.DAS.Reservations.Application/Reservations/Commands/CacheReservationEmployer/CacheReservationEmployerCommandValidator.cs (hunk 4 above)
Comment (robinoconnor) [reply src/SFA.DAS.Reservations.Application/Reservations/Commands/CacheReservationEmployer/CacheReservationEmployerCommandValidator.cs]: done

File: src/SFA.DAS.Reservations.Application/Reservations/Commands/CacheReservationStartDate/CacheReservationStartDateCommandValidator.cs (hunk 6 above)
Comment (robinoconnor) [reply src/SFA.DAS.Reservations.Application/Reservations/Commands/CacheReservationStartDate/CacheReservationStartDateCommandValidator.cs]: done

File: src/SFA.DAS.Reservations.Application/Reservations/Commands/CacheReservationStartDate/CacheReservationStartDateCommandHandler.cs (hunk 5 above)
Comment (robinoconnor) [reply src/SFA.DAS.Reservations.Application/Reservations/Commands/CacheReservationStartDate/CacheReservationStartDateCommandHandler.cs]: done

**Pull request merged**